# Future work

The current architecture is a proof of concept, with a concrete, then abstract
approach: have something working first with a public API, then refactor and
generalize. While it is deployed on AWS it can also be run in other cloud
providers, and the next goal is to replace most of the task queue and
communication with P2P technologies, using more of IPFS and dat (and not only as
file storage).

The WebAssembly support in sourmash also allows doing more data processing in
the browser, instead of transferring large datasets to the server. Currently
wort is more focused on the API and command line usage, but more functionality
will be added to the web frontend.

NCBI provides an alpha feature based on STAT to report the taxonomic composition
of reads within a sequencing run. This analysis can also be done with sourmash
gather, and a browser extension can overlay these results in the SRA Run
Browser. This extension idea is similar to what the BioJupies project [Torres,
2018] does for RNA-Seq datasets.

Any public database can be store and queried using wort, and we intend to add
more over time. 
